[Chin Cherng Yuen] iP

Arren11111 opened the pull request:
# FlyChat🪽 
>“Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes)

Flychat frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1.  Download the JAR file
2. Run the JAR file in your terminal using `java -jar flychat.jar` 
3. Add your tasks
4. And let it manage your tasks for you 👍 

**Features**
- [x] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders (_coming soon_)

---

```java
System.out.println("Manage your tasks well with FlyChat!")
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Arren11111 committed:
Changed Chatbot name, Implemented start and exit text

Arren11111 committed:
Changed start() and end() to private methods

Arren11111 committed:
Implemented the echo function

Arren11111 committed:
Implemented list adding and printing system

Arren11111 committed:
Added mark as done and undone ability

Arren11111 committed:
Added todo, event and deadline task types

Arren11111 committed:
Added I/O redirection testing functionality

Also implemented some minor bug fixes

Arren11111 committed:
Added Error Handling Capabilities

Arren11111 committed:
Added remove task functionality

Also added error handling for IndexOutOfBounds error for mark and unmark commands.

Arren11111 committed:
Add save function

Add a save function to the program

User task info is lost every time the program is closed, making the program unfriendly for normal use

Let's add a save function to allow users to save their tasks across multiple sessions

Arren11111 committed:
Merge branch 'branch-Level-7'

Arren11111 committed:
Save end date for deadline tasks as LocalDate objects

Arren11111 committed:
Merge branch 'branch-Level-8'

Arren11111 committed:
Break up code into more modular objects and place them in packages

Arren11111 committed:
Merge branch 'branch-More-OOP'

Arren11111 committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Arren11111 committed:
Add gradle support for FlyChat

Arren11111 committed:
Merge branch 'branch-A-Gradle'

Arren11111 committed:
Add Javadoc comments

Arren11111 committed:
Merge branch 'branch-A-JavaDoc'

emmannyyy commented on src/main/java/FlyChat/Core/FlyChat.java:
```diff
@@ -0,0 +1,97 @@
+package FlyChat.Core;
```
Hey, I like how organised this is! The package names are also very intuitive! 👍

emmannyyy commented on src/main/java/FlyChat/Core/FlyChat.java:
```diff
@@ -0,0 +1,97 @@
+        }
+    }
+
+    private static void processCommands() throws InputMismatchException, IOException {
```
Perhaps this can be replaced with a switch-case statement to make it neater?

emmannyyy commented on src/main/java/FlyChat/Core/Storage.java:
```diff
@@ -0,0 +1,68 @@
+                tmp.renameTo(saveFile);
+            }
+        } catch (IOException e) {
+            System.out.println("Unable to write to file. Information may not be saved for the next session.");
```
It's great that you gave this error message, I think the user will benefit knowing that the session won't be saved!

emmannyyy commented on src/main/java/FlyChat/Core/TaskList.java:
```diff
@@ -0,0 +1,116 @@
+        return "I've marked the task as not done T.T :\n  " + storageList.get(index - 1).toString();
+    }
+
+    public String deleteTask(int index) throws IndexOutOfBoundsException{
+        Task targetTask = storageList.get(index - 1);
+        storageList.remove(targetTask);
+        return "Task removed:\n  " + targetTask + "\nNow you have " + storageList.size() 
+                + " tasks in the list. SOLDIER ON! ^o^";
+    }
```
I really like the tone your chatbot has! Feels very friendly!

emmannyyy commented on src/main/java/FlyChat/Tasks/Deadline.java:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    @Override
+    public String formatStringForSaving() {
+        return "[D]" + super.toString() + " (by: " + this.deadline + ")";
+    }
```
Perhaps, you may want to add some JavaDoc comments to guide future users on what the format String in general should be.

emmannyyy commented on src/main/java/FlyChat/Tasks/Task.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Represents the Tasks of the user.
+ */
+public class Task {
```
Perhaps this class should be an Abstract task?
```suggestion
public abstract class Task {
```

emmannyyy commented on src/main/java/FlyChat/Tasks/Task.java:
```diff
@@ -0,0 +1,48 @@
+public class Task {
+    private boolean done;
+    private String text;
+
```
I noticed that the subclasses extending from Task all have an formatStringforSaving. 
Perhaps, for the Task Class, you can use a template string which acts as a Constant. That way, all the inheriting classes can get this formatString, allowing for code reuse :)

emmannyyy reviewed (commented):
Overall, I found your code neat, concise and easy to understand. Some things can be edited, such as using Switch-Case statements instead of multiple nested if-else statements.  I noted a few minor coding standard violations too, and some javadoc comments that were missing. 

Apart from that, LGTM! Just some nits to pick

Arren11111 committed:
Add JUnit tests

Arren11111 committed:
Merge branch 'branch-A-JUnit'

Arren11111 committed:
Amend save file file path

Arren11111 committed:
Adjust code to conform to coding standard

Arren11111 committed:
Add find command functionality

Arren11111 committed:
Merge branch 'branch-Level-9'

Arren11111 committed:
Merge branch 'branch-A-CodingStandard'

YangQF2002 commented on src/main/java/FlyChat/Core/FlyChat.java:
```diff
@@ -0,0 +1,99 @@
+package flychat.core;
```
This core package is quite nice currently, but I feel that it may need refactoring in future. 
For example, if theres alot more storage related files, likely need to wrap them into a separate Storage package instead!

YangQF2002 commented on src/main/java/FlyChat/Core/FlyChat.java:
```diff
@@ -0,0 +1,99 @@
+ */
+public class FlyChat {
+
+    private static Ui ui = new Ui();
+    private static Storage storage = new Storage();
+    private static TaskList taskList = new TaskList();
+    private static Parser parser = new Parser();
```
Should these be static final? Since the object is never reassigned.

YangQF2002 commented on src/main/java/FlyChat/Core/TaskList.java:
```diff
@@ -0,0 +1,186 @@
+        isCompleted = parser.checkTaskCompletedFromFile(str);
+
+        //format the task string properly using regex
+        if (parser.getTaskTypeFromFile(str).equals("[T]")) {
+            //To-do case
+            addToList(Todo.createNewTodo(parser.getTaskDescriptionFromFile(str), isCompleted));
+        } else if (parser.getTaskTypeFromFile(str).equals("[D]")) {
+            //Deadline case
+            addToList(Deadline.createNewDeadline(parser.getTaskDescriptionFromFile(str),
+                    parser.getDeadlineEndDateFromFile(str), isCompleted));
+        } else if (parser.getTaskTypeFromFile(str).equals("[E]")) {
```
Maybe `getTaskTypeFromFile` should just return `T`, `D`, or `E` without the brackets. 
Just a minor nitpick.

YangQF2002 commented on src/main/java/FlyChat/Core/Ui.java:
Maybe consider abstracting out the double linebreak print into a separate method? To prevent code repetition.

YangQF2002 reviewed (commented):
Generally, LGTM. Just some minor things to take note of. Good job!!

Arren11111 committed:
Add JavaFX GUI

Arren11111 committed:
Add assertions to code

Add assertions to classes

Unanticipated behaviors may break some assumptions regarding code. This may lead to runtime errors if left unchecked.

Let's add assertions to the code.

Assertions ensure that vital code assumptions are met and checked during compile time.

Arren11111 committed:
Improve code quality

Code quality in some areas of the program can be improved.

Ensuring good code quality across the program can make it much easier to debug and understand.

Extracting the long parseCommand method and addTask method into different Command classes make the code much more legible. Tweaking the program to fit coding conventions also helps to maintain consistent style across the program, enhancing legibility.

Let's extract out the parseCommand and addTask methods into Command classes using enumeration and make tweaks to the program to follow the course coding standard.

Arren11111 committed:
Merge pull request #2 from Arren11111/branch-A-CodeQuality

Merge branch-A-CodeQuality

Arren11111 committed:
Merge remote-tracking branch 'origin/master'

Arren11111 committed:
Add tagging functionality

Tasks have no way of being grouped according to similarities or relations to one another.

Having a grouping system for related tags enhances functionality for the user and enables more extensive organisation.

Let's make it possible for users to tag their tasks according to their relation to each other

Tags allow users to define how they want to group their tasks together and serve as an effective yet simple way of organising tasks.

Arren11111 committed:
Merge branch 'branch-C-Tagging'

Arren11111 committed:
Fix goodbye message unable to be displayed

Arren11111 committed:
Simplify comments

Arren11111 committed:
Edit README.md

Arren11111 committed:
Edit README.md

Arren11111 committed:
Edit README.md

Arren11111 committed:
Edit README.md

Arren11111 committed:
Fix bugs in Parser.java
